### PR TITLE
httpprovider for Collector: load configuration from config files in HTTP servers

### PR DESCRIPTION
**Pull request by junhaoyu-aws**
**Description:** 
An implementation of ConfigMapProvider for HTTP (httpprovider) allows OTEL Collector the ability to load configuration for itself by fetching and reading config files stored in HTTP servers. 

What I changed: added one more new provider implementation folder under confmap/provider/, also added httpprovider.New() in the initialization of a default ConfigProviderSettings.

**Link to tracking Issue:** <Issue number if applicable>
#5810 

**Testing:** 
First, I mocked how a HTTP client works while interacting with a HTTP server.
Second, I designed and passed all the unit tests for edge cases such as empty fields, invalid config files, non-exist config files and so on.
Third, I did integration tests such as:
- Start an OTEL Collector with httpmapprovider.
- Start an OTEL sample application to send telemetry data to check if the OTEL Collector is running well.

**Documentation:** 
Added a README.md

**Sponsor:** 
@Aneurysm9

### CODE REVIEW

File: confmap/provider/httpprovider/provider.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+// One example for http-uri be like: http://localhost:3333/getConfig
+//
+// Examples:
+// `http://localhost:3333/getConfig` - (unix, windows)
+func New() confmap.Provider {
+	return &provider{client: &http.Client{}}
+}
+
+func (fmp *provider) Retrieve(_ context.Context, uri string, _ confmap.WatcherFunc) (*confmap.Retrieved, error) {
+	if !strings.HasPrefix(uri, schemeName+"://") {
```
Comment (bogdandrutu) [inline confmap/provider/httpprovider/provider.go:53]: why do you need the authority (I think that is between the `/`) to be empty?

File: confmap/provider/httpprovider/provider.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	"go.opentelemetry.io/collector/confmap/provider/internal"
+)
+
+const (
+	schemeName = "http"
+)
+
+type httpClient interface {
+	Get(url string) (resp *http.Response, err error)
+}
```
Comment (bogdandrutu) [inline confmap/provider/httpprovider/provider.go:34]: Instead of creating a "client", better test against an http test server, see https://pkg.go.dev/net/http/httptest#NewServer

File: confmap/provider/httpprovider/provider_test.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
+// Copyright The OpenTelemetry Authors
```
Comment (bogdandrutu) [inline confmap/provider/httpprovider/provider_test.go:1]: Add a test like https://github.com/open-telemetry/opentelemetry-collector/blob/main/confmap/provider/fileprovider/provider_test.go#L32

**Review by bogdandrutu — COMMENTED** [review]
Please update the PR title and commit message accordingly, even though you may upstream this, we need to use nicer commit messages

File: confmap/provider/httpprovider/provider.go (hunk 1 above)
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/provider.go]: Fixed, you are right. I took a look at fileprovider.

File: confmap/provider/httpprovider/provider.go (hunk 2 above)
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/provider.go:34]: Fixed, that makes sense.

File: confmap/provider/httpprovider/provider_test.go (hunk 3 above)
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/provider_test.go:1]: Added.

File: confmap/provider/httpprovider/README.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+What is this new component httpprovider?
+- An implementation of ConfigMapProvider for HTTP (httpprovider) allows OTEL Collector the ability to load configuration for itself by fetching and reading config files stored in HTTP servers.
```
Comment (bogdandrutu) [inline confmap/provider/httpprovider/README.md:2]: ConfigMapProvider is the old name please update everywhere

File: confmap/provider/httpprovider/README.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+How this new component httpprovider works?
+- It will be called by ConfigMapResolver to load configurations for OTEL Collector.
```
Comment (bogdandrutu) [inline confmap/provider/httpprovider/README.md:5]: The Resolver is in a different package now

File: confmap/provider/httpprovider/provider.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httpprovider // import "go.opentelemetry.io/collector/confmap/provider/httpprovider"
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
```
Comment (bogdandrutu) [inline confmap/provider/httpprovider/provider.go:20]: Please don't use deprecated packages

File: confmap/provider/httpprovider/provider.go (hunk 2 above)
Comment (bogdandrutu) [reply confmap/provider/httpprovider/provider.go:34]: Doesn't seem to be fixed

File: confmap/provider/httpprovider/provider_test.go (hunk 3 above)
Comment (bogdandrutu) [reply confmap/provider/httpprovider/provider_test.go:1]: Did you forget to push the commit?

File: confmap/provider/httpprovider/provider.go (hunk 2 above)
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/provider.go:34]: Oh, I updated the unit test with httptest package. Do I also need to update provider.go?

File: confmap/provider/httpprovider/provider_test.go (hunk 3 above)
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/provider_test.go:1]: I think added at line 138 in provider_test.go?

File: confmap/provider/httpprovider/README.md (hunk 4 above)
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/README.md]: Fixed!

File: confmap/provider/httpprovider/provider.go (hunk 6 above)
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/provider.go]: Fixed

File: confmap/provider/httpprovider/provider_test.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+	fp := NewTestProvider()
+	_, err := fp.Retrieve(context.Background(), "https://...", nil)
+	assert.Error(t, err)
+	assert.NoError(t, fp.Shutdown(context.Background()))
+}
+
+func TestEmptyURI(t *testing.T) {
+	fp := NewTestProvider()
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(404)
```
Comment (bogdandrutu) [inline confmap/provider/httpprovider/provider_test.go:81]: Return 200 if this is "empty"

File: confmap/provider/httpprovider/provider_test.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+	require.NoError(t, fp.Shutdown(context.Background()))
+}
+
+func TestNonExistent(t *testing.T) {
+	fp := NewTestNonExistProvider()
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		f, err := os.ReadFile("./testdata/nonexist-otel-config.yaml")
+		if err != nil {
+			w.WriteHeader(404)
+			_, innerErr := w.Write([]byte("Cannot find the config file"))
```
Comment (bogdandrutu) [inline confmap/provider/httpprovider/provider_test.go:99]: don't write anything else, just 404 is enough.

File: confmap/provider/httpprovider/provider_test.go [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+			fmt.Println("Write failed: ", err)
+		}
+	}))
+	defer ts.Close()
+	_, err := fp.Retrieve(context.Background(), ts.URL, nil)
+	require.Error(t, err)
+	require.NoError(t, fp.Shutdown(context.Background()))
+}
+
+func TestNonExistent(t *testing.T) {
```
Comment (bogdandrutu) [inline confmap/provider/httpprovider/provider_test.go:93]: Add a new test that tries to retrieve from an already shutdown server.

File: confmap/provider/httpprovider/provider.go (hunk 2 above)
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/provider.go]: Fixed!

File: confmap/provider/httpprovider/README.md [hunk 10]
```diff
@@ -0,0 +1,13 @@
+What is this new component httpprovider?
+- An implementation of ConfigProvider for HTTP (httpprovider) allows OTEL Collector the ability to load configuration for itself by fetching and reading config files stored in HTTP servers.
```
[the remark this answers is not in the document]
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/README.md]: Fixed!

File: confmap/provider/httpprovider/README.md [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+
+How this new component httpprovider works?
+- It will be called by ConfigMapResolver to load configurations for OTEL Collector.
```
[the remark this answers is not in the document]
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/README.md]: Fixed!

File: confmap/provider/httpprovider/provider_test.go [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"go.opentelemetry.io/collector/confmap"
+	"go.opentelemetry.io/collector/confmap/confmaptest"
+)
+
+// Create a provider mocking httpmapprovider works in normal cases
+func NewTestProvider() confmap.Provider {
```
[the remark this answers is not in the document]
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/provider_test.go]: Fixed!

File: confmap/provider/httpprovider/provider_test.go [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+	return &provider{client: &http.Client{}}
+}
+
+// Create a provider mocking httpmapprovider works when the returned config file is invalid
+func NewTestInvalidProvider() confmap.Provider {
```
[the remark this answers is not in the document]
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/provider_test.go]: Fixed!

File: confmap/provider/httpprovider/provider_test.go [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+	return &provider{client: &http.Client{}}
+}
+
+// Create a provider mocking httpmapprovider works when there is no corresponding config file according to the given http-uri
+func NewTestNonExistProvider() confmap.Provider {
```
[the remark this answers is not in the document]
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/provider_test.go]: Fixed!

File: confmap/provider/httpprovider/provider_test.go (hunk 7 above)
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/provider_test.go]: Hi Bogdan, I am still a little bit confused. I think it should return 400 instead? The case is the HTTP URI is empty, not the content of config file is empty?

File: confmap/provider/httpprovider/provider_test.go (hunk 8 above)
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/provider_test.go]: Fixed!

File: confmap/provider/httpprovider/provider_test.go (hunk 9 above)
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/provider_test.go:84]: Added!

File: confmap/provider/httpprovider/provider_test.go [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+		f, err := os.ReadFile("./testdata/nonexist-otel-config.yaml")
+		if err != nil {
+			w.WriteHeader(404)
+			return
+		}
+		w.WriteHeader(200)
+		_, err = w.Write(f)
+		if err != nil {
+			fmt.Println("Write failed: ", err)
+		}
```
Comment (bogdandrutu) [inline confmap/provider/httpprovider/provider_test.go:96]:
```suggestion
		w.WriteHeader(404)
```

File: confmap/provider/httpprovider/provider_test.go [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+	require.Error(t, err)
+	require.NoError(t, fp.Shutdown(context.Background()))
+}
+
+func TestRetrieveFromShutdownServer(t *testing.T) {
+	fp := New()
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(404)
+	}))
+	defer ts.Close()
```
Comment (bogdandrutu) [inline confmap/provider/httpprovider/provider_test.go:78]:
```suggestion
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
	ts.Close()
```

File: confmap/provider/httpprovider/provider_test.go (hunk 15 above)
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/provider_test.go]: Fixed.

File: confmap/provider/httpprovider/provider_test.go (hunk 16 above)
Comment (junhaoyu-aws) [reply confmap/provider/httpprovider/provider_test.go]: Fixed.

**Review by bogdandrutu — APPROVED** [review]

File: confmap/provider/httpprovider/provider.go [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+// New returns a new confmap.Provider that reads the configuration from a file.
+//
+// This Provider supports "http" scheme, and can be called with a "uri" that follows:
+//
+// One example for http-uri be like: http://localhost:3333/getConfig
+//
+// Examples:
+// `http://localhost:3333/getConfig` - (unix, windows)
+func New() confmap.Provider {
+	return &provider{client: http.Client{}}
```
Comment (josephwoodward) [inline confmap/provider/httpprovider/provider.go:45]: Given this can't be controlled by the user, would it be sensible to set a default timeout value as opposed to the client's default value of 0?

Comment (bogdandrutu) [reply confmap/provider/httpprovider/provider.go:45]: I think we should do that in the context we pass to "Retrieve" from the service package. So that we have it across all the providers?
